### PR TITLE
Add authentication and nicknames for comments

### PR DESCRIPTION
**Pull request by n-wach**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.User;
+
+/**
+ * Handles getting user information from the datastore.  This enables
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:28]: Looks like the comment is incomplete?
Also for new paragraphs use <p> with a new line before it.

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+public class Authenticator {
+  private static Authenticator instance = new Authenticator();
+
+  private UserService userService;
+  private DatastoreService datastore;
+
+  private Authenticator() {
+    userService = UserServiceFactory.getUserService();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:39]: Its best practice to inject this like so.
private Authenticator(UserService userService, DatastoreService datastore){
...}

This will dramatically help with testing as you will be able to mock these dependencies instead of needing to instantiate them in tests. A class that favors injection is usually much more testable.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:39]: Ok, since it's a singleton should I instantiate with `private static Authenticator instance = new Authenticator(UserServiceFactory.getUserService(), DatastoreServiceFactory.getDatastoreService());`? Or instantiate like this in .getAuthenticator if instance is null?

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:39]: Maybe also overload .getAuthenticator to be able to specify specific userService/datastore ?

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.User;
+
+/**
+ * Handles getting user information from the datastore.  This enables
+ */
+public class Authenticator {
+  private static Authenticator instance = new Authenticator();
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:31]: Is this supposed to be final?

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  private UserService userService;
+  private DatastoreService datastore;
+
+  private Authenticator() {
+    userService = UserServiceFactory.getUserService();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public static Authenticator getAuthenticator() {
+    return instance;
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:42]:
```suggestion
    return INSTANCE;
```

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      return false;
+    }
+
+    return true;
+  }
+
+  private void setNickname(String nickname) {
+    String id = auth.getCurrentUser().id();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:131]: can you resuse the initialized datastore?

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java]: Done

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 1 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:28]: Done

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java]: Fixed

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public static Authenticator getAuthenticator() {
+    return INSTANCE;
+  }
+
+  /**
+   * Returns the {@code User} object associated with the current user, or a new
+   * User with the empty string as a nickname if one cannot be found.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:50]:
```suggestion
   * {@code User} with the empty string as a nickname if one cannot be found.
```

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+   */
+  public User getCurrentUser() {
+    String id = userService.getCurrentUser().getUserId();
+    Query query = new Query(User.ENTITY_KIND)
+        .setFilter(new Query.FilterPredicate(User.PROP_ID, Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return User.create(id, "", false);
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:59]: Please add inline comments for literal arguments.

"Document ambiguous literal arguments at the call site
In certain contexts literal arguments - such as 0, "", true and false, or null - can make it difficult for readers to know what a method will do. We generally discourage defining methods that take boolean parameters or otherwise expect users to pass in ambiguous literals. However, when you must call such a method you're encouraged to use the parameter name as an inline comment at the call site, so that readers don't need to look at the method declaration to understand the parameter's purpose.

Such a comment should be formatted in this way:

booleanMethod(/* enableFoo= */ true);"

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    }
+    return User.fromEntity(entity);
+  }
+
+  /**
+   * Returns if the current user is logged in.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:65]:
```suggestion
   * Returns {@code true} if the current user is logged in.
```

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 9]
```diff
@@ -22,25 +22,34 @@
...
+    String author = (String) entity.getProperty(PROP_AUTHOR);
     String topic = (String) entity.getProperty(PROP_TOPIC);
     String content = (String) entity.getProperty(PROP_CONTENT);
     Date created = (Date) entity.getProperty(PROP_CREATED);
-    return Comment.create(key, topic, content, created);
+    return Comment.create(key, author, topic, content, created);
   }
 
-  /** Returns the comment's Datastore entity key. */
+  /** Returns the comment's Datastore entity key */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:44]: Why did you remove the period? 

https://developers.google.com/style/api-reference-comments#classes_interfaces_structs

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.auto.value.AutoValue;
+import java.util.Optional;
+
+/** Contains login status information. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/LoginStatus.java:20]: Maybe instead of "Contains", it should be "Stores"?

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+@AutoValue
+public abstract class LoginStatus {
+
+  /**
+   * Returns the currently logged in User associated with this login. If not
+   * logged in, returns {@code Optional.empty()}.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/LoginStatus.java:26]: I would re-phrase the second sentence as: If no user is logged in, returns {@code Optional.empty()}

File: portfolio/src/main/java/com/google/sps/data/User.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.auto.value.AutoValue;
+
+/** Contains user information. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/User.java:20]: I think you can omit "Contains" as it is an auto value class.

File: portfolio/src/main/java/com/google/sps/data/User.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  /** Returns the user's id as given by {@code UserService} */
+  public abstract String id();
+
+  /** Returns the user's nickname. */
+  public abstract String nickname();
+
+  /**
+   * Returns if the user has admin priviledges.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/User.java:46]:
```suggestion
   * Returns {@code true} if the user has admin privileges.
```

File: portfolio/src/main/java/com/google/sps/data/User.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+   *
+   * <p>Being an admin is required for certain actions. It defaults to false,
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/User.java:48]: Can you specify what actions admin is required for?

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 15]
```diff
@@ -44,7 +45,7 @@
 /**
  * Servlet that saves and serves comments for Nathan's site.
  *
- * <p>Storage is done with the GCP Datastore Service.
+ * <p>Storage is done with the GCP Datastore Service
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:48]: Why did you remove a period?

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+import com.google.sps.data.LoginStatus;
+import com.google.sps.data.User;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/**
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:35]: This comment has wrong spacing. It should be 

/**
 *
 */

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+      } else if (!isValidNickname(nickname)) {
+        response.sendError(400, "Invalid nickname");
+        return;
+      } else {
+        setNickname(nickname);
+        return;
+      }
+    }
+    response.sendError(400, "Nothing to do");
+    return;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:81]: This entire if statement can be simplified to this:

if (nickname == null) {
 response.sendError(400, "Nothing to do");
  return;
}

 if (isValidNickname(currentNickname)) {
   response.sendError(400, "A valid nickname may not be changed");
   return;
}

if (!isValidNickname(nickname)) {
  response.sendError(400, "Invalid nickname");
  return;
}

setNickname(nickname);
return;

Do you see how it is simpler to reason about?

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    response.setContentType("application/json;");
+    response.getWriter().println(threadJson);
+  }
+
+  private boolean isValidNickname(String nickname) {
+    if (nickname.length() < MIN_NICKNAME_LEN || nickname.length() > MAX_NICKNAME_LEN) {
+      return false;
+    }
+    if (nickname.toLowerCase().equals(NICKNAME_ANONYMOUS)) {
+      // Reserved
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:111]: What does reserved mean? Can you expand on the comment?

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -137,11 +197,23 @@ function makeSubmission(topic) {
 function inflateThread(threadElement) {
   const topic = threadElement.getAttribute(ATTR_TOPIC);
 
-  const submissionForm = makeSubmission(topic);
-  threadElement.appendChild(submissionForm);
+  if (USER_INFO.loggedIn) {
+    // User can only submit comments if they are logged in.
+    // And have a nickname
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:202]: "And have a nickname" is not a proper sentence. Can you rephrase?

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -137,11 +197,23 @@ function makeSubmission(topic) {
...
+    if (USER_INFO.user.value.nickname !== "") {
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:203]: To make these nested for loops less complicated I would do either:

if (USER_INFO.loggedIn && USER_INFO.loggedIn) {
} else if (USER_INFO.loggedIn) {
} else {
}

or I would do:
if (USER_INFO.loggedIn) {
  // call method 
} else {
}

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:50]: Fixed

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 7 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:59]: Added

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 8 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:65]: Added

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 9 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:44]: lol no idea

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:44]: fixed

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java (hunk 11 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/LoginStatus.java:26]: Done

File: portfolio/src/main/java/com/google/sps/data/User.java (hunk 13 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/User.java:46]: Fixed

File: portfolio/src/main/java/com/google/sps/data/User.java (hunk 14 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/User.java:48]: I can, but I assume that'd mean I'd want to keep it up to date as I add new features... seem's like it'd be hard to remember to update it

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 15 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:48]: I have no idea where they're going lol.  Fixed

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 16 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:35]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 17 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:81]: Yeah.  I wanted this to be able to take more than just `nickname`.  I turned it into this:

```
...
    if (nickname != null) {
      if (isValidNickname(currentNickname)) {
        response.sendError(400, "A valid nickname may not be changed");
        return;
      }

      if (!isValidNickname(nickname)) {
        response.sendError(400, "Invalid nickname");
        return;
      }

      setNickname(nickname);
      return;
    }
...
```

Would have other null checks for other params as I add them... would it be better to hold off on that until I need it?

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 18 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:111]: Done

File: portfolio/src/main/webapp/script.js (hunk 19 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:202]: Cleaned this up

File: portfolio/src/main/webapp/script.js (hunk 20 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:203]: Cleaned this up

File: portfolio/src/main/java/com/google/sps/data/User.java (hunk 12 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/User.java:20]: Keeping as all other auto value data classes use "Contains"

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java (hunk 10 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/LoginStatus.java:20]: Keeping as all other auto value data classes use "Contains"

**Comment by n-wach** [comment]
Submission field when logged in + test comment: https://screenshot.googleplex.com/THXm9gEELXx

Login field: https://screenshot.googleplex.com/d1JK4dQcLLa

Nickname field: https://screenshot.googleplex.com/pqYLRd3qzEH

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+  private Authenticator() {
+    userService = UserServiceFactory.getUserService();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public static Authenticator getAuthenticator() {
+    return INSTANCE;
+  }
+
+  /**
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:48]: nit: consider returning optional.empty() for no user.

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:48]: Optional is more obvious and less bug prone. The user of this method needs to verify it isn't empty to use the result.

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+   * {@code User} with the empty string as a nickname if one cannot be found.
+   */
+  public User getCurrentUser() {
+    String id = userService.getCurrentUser().getUserId();
+    Query query = new Query(User.ENTITY_KIND)
+        .setFilter(new Query.FilterPredicate(User.PROP_ID, Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return User.create(id, /** nickname= */ "", userService.isUserAdmin());
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/auth/Authenticator.java:59]: Right now you are using a java doc comment. For this you need to use the regular comment /* */.

File: portfolio/src/main/java/com/google/sps/data/User.java (hunk 14 above)
Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/data/User.java:48]: makes sense

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 17 above)
Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:84]: I would first check if nickname is null and return early. Once you need to add more variables we can look into it. I can imagine with other variables you could do:

if (nickname == null || bla == null | bla == null) {
  // return early
}

But we can worry about it later.

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    if (nickname.length() < MIN_NICKNAME_LEN || nickname.length() > MAX_NICKNAME_LEN) {
+      return false;
+    }
+
+    // Nickname cannot be "anonymous", as this is reserved for anonymous comments.
+    if (nickname.toLowerCase().equals(NICKNAME_ANONYMOUS)) {
+      return false;
+    }
+
+    // Nickname must be unique
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:119]:
```suggestion
    // Nickname must be unique.
```

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 17 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:84]: Sounds good, changed it

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 21 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:48]: There will always be a user, it's just that there might not always be a database entity for that user.  If there isn't one, we want to create it.  I figure it's easier to create it here then in every possible place where .getCurrentUser might be called

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 22 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:59]: ah, fixed

File: portfolio/src/main/java/com/google/sps/auth/Authenticator.java (hunk 21 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/auth/Authenticator.java:48]: Added a comment clarifying

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java (hunk 23 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:119]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+      return;
+    }
+
+    if (!isValidNickname(nickname)) {
+      response.sendError(400, "Invalid nickname");
+      return;
+    }
+
+    setNickname(nickname);
+    return;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:85]: nit: this can be omitted given you are returning void and you are at the end of the method.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java:85]: Ah yes

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UsersServlet.java]: Fixed
